File: SourceView/FileViewController.m
/*
 Copyright (C) 2018 Apple Inc. All Rights Reserved.
 See LICENSE.txt for this sample’s licensing information
 
 Abstract:
 View controller object to host the UI for file information
 */

#import "FileViewController.h"

@interface FileViewController ()

@property (nonatomic, weak) IBOutlet NSImageView *fileIcon;
@property (nonatomic, weak) IBOutlet NSTextField *fileName;
@property (nonatomic, weak) IBOutlet NSTextField *fileSize;
@property (nonatomic, weak) IBOutlet NSTextField *modDate;
@property (nonatomic, weak) IBOutlet NSTextField *creationDate;
@property (nonatomic, weak) IBOutlet NSTextField *fileKindString;

@end

#pragma mark -

@implementation FileViewController

// -------------------------------------------------------------------------------
//	awakeFromNib
// -------------------------------------------------------------------------------
- (void)awakeFromNib
{
	// listen for changes in the url for this view
	[self addObserver:	self
						forKeyPath:@"url"
						options:(NSKeyValueObservingOptionNew | NSKeyValueObservingOptionOld) 
						context:nil];
}

// -------------------------------------------------------------------------------
//	dealloc
// -------------------------------------------------------------------------------
- (void)dealloc
{
	[self removeObserver:self forKeyPath:@"url"];
}

// -------------------------------------------------------------------------------
//	observeValueForKeyPath:ofObject:change:context
//
//	Listen for changes in the file url.
// -------------------------------------------------------------------------------
- (void)observeValueForKeyPath:	(NSString *)keyPath
								ofObject:(id)object 
								change:(NSDictionary *)change 
								context:(void *)context
{
	// name
	self.fileName.stringValue = [[NSFileManager defaultManager] displayNameAtPath:self.url.path];
	
	// icon
	NSImage *iconImage = [[NSWorkspace sharedWorkspace] iconForFile:self.url.path];
	iconImage.size = NSMakeSize(64,64);
	self.fileIcon.image = iconImage;
	
	NSDictionary *attr = [[NSFileManager defaultManager] attributesOfItemAtPath:self.url.path error:nil];
	if (attr)
	{
		// file size
		NSNumber *theFileSize = attr[NSFileSize];
        self.fileSize.stringValue = [NSString stringWithFormat:@"%@ KB on disk", theFileSize.stringValue];
		
		// creation date
		NSDate *fileCreationDate = attr[NSFileCreationDate];
        self.creationDate.stringValue = fileCreationDate.description;
				
		// mod date
		NSDate *fileModDate = attr[NSFileModificationDate];
        self.modDate.stringValue = fileModDate.description;
	}

	// kind string
	NSString *kindStr;
    [self.url getResourceValue:&kindStr forKey:NSURLLocalizedTypeDescriptionKey error:nil];
	if (kindStr != nil)
	{
		self.fileKindString.stringValue = kindStr;
	}
}

@end

File: SourceView/FileViewController.swift
//
//  FileViewController.swift
//  SourceView
//
//  Translated by OOPer in cooperation with shlab.jp, on 2015/4/6.
//
//
/*
 Copyright (C) 2018 Apple Inc. All Rights Reserved.
 See LICENSE.txt for this sample’s licensing information

 Abstract:
 View controller object to host the UI for file information
 */

import Cocoa

@objc(FileViewController)
class FileViewController: NSViewController {
    
    var url: URL?
    
    @IBOutlet private var fileIcon: NSImageView!
    @IBOutlet private var fileName: NSTextField!
    @IBOutlet private var fileSize: NSTextField!
    @IBOutlet private var modDate: NSTextField!
    @IBOutlet private var creationDate: NSTextField!
    @IBOutlet private var fileKindString: NSTextField!
    
    //MARK: -
    
    // -------------------------------------------------------------------------------
    //	awakeFromNib
    // -------------------------------------------------------------------------------
    override func awakeFromNib() {
        // listen for changes in the url for this view
        self.addObserver(self,
            forKeyPath: "url",
            options: [.new, .old],
            context: nil)
    }
    
    // -------------------------------------------------------------------------------
    //	dealloc
    // -------------------------------------------------------------------------------
    deinit {
        self.removeObserver(self, forKeyPath: "url")
    }
    
    // -------------------------------------------------------------------------------
    //	observeValueForKeyPath:ofObject:change:context
    //
    //	Listen for changes in the file url.
    // -------------------------------------------------------------------------------
    override func observeValue(forKeyPath keyPath: String?,
        of object: Any?,
        change: [NSKeyValueChangeKey : Any]?,
        context: UnsafeMutableRawPointer?)
    {
        if let url = self.url {
            let path = url.path
            // name
            self.fileName.stringValue = FileManager.default.displayName(atPath: path)
            
            // icon
            let iconImage = NSWorkspace.shared().icon(forFile: path)
            iconImage.size = NSMakeSize(64, 64)
            self.fileIcon.image = iconImage
            if let attr = try? FileManager.default.attributesOfItem(atPath: path) {
                // file size
                let theFileSize = attr[FileAttributeKey.size] as! NSNumber
                self.fileSize.stringValue = "\(theFileSize.stringValue) KB on disk"
                
                // creation date
                let fileCreationDate = attr[FileAttributeKey.creationDate] as! Date
                self.creationDate.stringValue = fileCreationDate.description
                
                // mod date
                let fileModDate = attr[FileAttributeKey.modificationDate] as! Date
                self.modDate.stringValue = fileModDate.description
            }
            
            // kind string
            let resource = try? url.resourceValues(forKeys: [.localizedTypeDescriptionKey])
            let kindStr = resource?.localizedTypeDescription
            if let str = kindStr {
                self.fileKindString.stringValue = str
            }
        } else {
            self.fileName.stringValue = ""
            self.fileIcon.image = nil
            self.fileSize.stringValue = ""
            self.creationDate.stringValue = ""
            self.modDate.stringValue = ""
            self.fileKindString.stringValue = ""
        }
    }
    
}
